### PR TITLE
SPI fixes

### PR DESCRIPTION
**Pull request by mbolivar**
This series makes improvements to the driver core for, and fixes bugs in, new-style SPI drivers.

- Add status reporting to the SPI driver core's synchronous calls to enable better error handling and reporting
- Add SPI support to 96b_carbon (STM32F4 based)
- Make various fixes to the STM32 SPI driver

Without this series, SPI master on STM32F4 is not working in my testing. With it, the loopback tests pass on 96b_carbon, and I've also verified that the behavior of other protocol features (like NSS pins) not tested in loopback are working correctly using a logic analyzer.

I don't however have an STM32F3 or other board with the FIFO-based IP block to test on -- @erwango, @superna9999, @ldts, any testing here would be much appreciated.

Review requested from @tbursztyka as well.

### CODE REVIEW

**Comment by erwango** [comment]
Hi @mbolivar,
I had some quick tests. not much time to analyze more for now:
*nucleo_l476rg: test is stalled (irq & poll)
   > [general] [INF] spi_complete_loop: Start ..... Then hangs
   > in spi_stm32_shift_m, waiting for RXNE flag
   > regression vs previous status
*nucleo_f334r8: bus error: (irq & poll)
  Faulting instruction address:  0x8001f0e  (drivers/spi/spi_context.h:229)
  Imprecise data bus error
Fatal fault in ISR! Spinning...
   > Not a regression (hum... this is embarrassing :/)

Hope it helps.

**Comment by mbolivar** [comment]
> I had some quick tests. not much time to analyze more for now:

Awesome, thanks for the testing!

So something is wrong with this implementation in the case of SPI peripherals with FIFOs (F3 and L4). It looks like we'll need to select a spi_stm32_shift_frame() implementation depending on the STM32 family's SPI peripheral type.

It looks like this will be easy to do if we check for the presence of the FIFO threshold bit in CR2:

```
#ifdef SPI_CR2_FRXTH
/* FIFO spi_stm32_shift_frame() implementation goes here */
#else
/* non-FIFO spi_stm32_shift_frame() implementation goes here */
#endif
```

Seem OK and portable across STM32 SoCs?

(I would really prefer to avoid having to add in a CONFIG_SPI_STM32_VERSION that we then fill into the Kconfig for every SoC, or, worse, board defconfig when the information we need is available to us from CMSIS already. I.e. take a different approach than the STM32 I2C "V1" and "V2" configuration mechanism.)

**Comment by mbolivar** [comment]
@erwango I've rebased and updated this per our discussion in email. I've tried to keep the existing behavior unchanged in the FIFO case. If you have some time to retest, I'd appreciate it. Tests are passing on STM32F4.

**Comment by erwango** [comment]
@mbolivar, test goes a little bit further on L476:
`�[general] [INF] main: SPI test on buffers TX/RX 0x2000000c/0x20000274
[general] [DBG] spi_async_call_cb: Polling...
[general] [INF] spi_complete_loop: Start
[general] [DBG] spi_stm32_configure: Installed config 0x20000038: freq 312500Hz (div 0
[general] [DBG] spi_context_buffers_setup: tx_bufs 0x20000b50 (1) - rx_bufs 0x20000b51
[general] [DBG] spi_context_buffers_setup: current_tx 0x20000b50 (1), current_rx 0x207
[general] [DBG] spi_context_update_tx: tx buf/len 0x2000000d/16`
But stall and remains stuck at:
`0x08001d00 in spi_stm32_shift_m (data=0x200000c0 <spi_stm32_dev_data_1>, spi=0x40013000) at /local/mcu/zephyr/zephyr-project/drivers/spi/spi_ll_stm32.c:68
68		while (!LL_SPI_IsActiveFlag_RXNE(spi)) {`

F334r8 has same status as previously. For this one, since we know this is not a regression and
this is not linked with your dev, I'll raise a JIRA (or github issue, or..)

**Comment by mbolivar** [comment]
> But stall and remains stuck at:

Hrm, I wasn't expecting that :(.

I'll wait to get some L4 hardware to test on here and update this PR when I've had a chance to look into this further.

**Comment by mbolivar** [comment]
Thanks to @erwango 's kindly sending me an L4-based board, I was able to retest this on that SPI IP. I've rebased and updated the series to include some fixes for STM32L4 as well, which is also currently broken on mainline, and the problems with which I'm grateful to Erwan for diagnosing.

I've verified spi_loopback functionality with this series on STM32L4 and STM32F4 in both polled and loopback modes.

@tbursztyka @erwango -- could you please review?

**Comment by mbolivar** [comment]
Not sure why shippable is failing after I added the CODEOWNERS commit; any ideas? Looks like a flaky test.

**Comment by ydamigos** [comment]
@mbolivar I got the same error yesterday. Anas informed me that it wasn't my fault.

**Review by erwango — APPROVED** [review]
LGTM
As mentioned in comments:
*Tested ok on L4
*Issue on F3 which does not look related to this fix. ZEP-2419 raised for this purpose.

**Comment by mbolivar** [comment]
Thanks, Erwan.

@tbursztyka, I'm especially hoping for your OK on de50763, which affects the driver core.

**Comment by mbolivar** [comment]
Rebased and dropped the first two patches from the previous version, which were made obsolete by #888. Tests still pass on F4 and L4.

**Comment by superna9999** [comment]
Sorry I was in vacation, I'll get a run of the patchset.

First of all, I don't really understand why you treat master and slave shifting differently.
From a pure SPI bus PoV, it's totally identical, did you run into an issue somewhere ?

File: CODEOWNERS
Comment (superna9999) [inline CODEOWNERS]: please add the other contributors like @ldts and me

File: drivers/spi/spi_ll_stm32.c
Comment (superna9999) [inline drivers/spi/spi_ll_stm32.c]: Maybe you can force it as inline

File: drivers/spi/spi_ll_stm32.c
Comment (superna9999) [inline drivers/spi/spi_ll_stm32.c]: Same here

File: drivers/spi/spi_ll_stm32.c
Comment (superna9999) [inline drivers/spi/spi_ll_stm32.c]: I don't get the true difference apart that in the master shift you consider you want to push a byte and then get a byte, but why is it different for slave and why can't we leave the previous way for both ?

File: CODEOWNERS
Comment (mbolivar) [inline CODEOWNERS]: Do you want to add yourselves? There's no issue in doing so -- I just didn't want to "volunteer" you for future review.

File: drivers/spi/spi_ll_stm32.c
Comment (mbolivar) [inline drivers/spi/spi_ll_stm32.c]: I want to let GCC decide. Any issues?

File: drivers/spi/spi_ll_stm32.c
Comment (mbolivar) [inline drivers/spi/spi_ll_stm32.c]: Same; I like to let the compiler decide.

File: drivers/spi/spi_ll_stm32.c
Comment (mbolivar) [inline drivers/spi/spi_ll_stm32.c]: Since the master is in charge of the bus and controls when frames get shifted, it's able to block until the TX/RX registers are empty/full respectively to move the I/O along. But doing that from the slave felt like asking for trouble, so I made it more of a "best effort" basis.

If by "previous way" you mean the code as is, I tried to explain in my commit logs why it's not correct: to transmit, it writes to DR, then doesn't always read back the frame that got shifted in by waiting until DR has data to read and pulling it out, freeing up the RX register for the next time a frame needs to be clocked out. This is causing overrun errors (OVR bit in SR).

**Comment by mbolivar** [comment]
@superna9999 are you OK with my answers to your questions?

**Comment by superna9999** [comment]
@mbolivar Yes, please add me to the code owners, for the other questions, I'm still doubious since it's more an issue on non-FIFO devices, but it will keep code simple, I will live fine with it.

So LGTM

**Comment by mbolivar** [comment]
@superna9999 I totally agree with you that the behavior in this series is suboptimal for the FIFO case. My goal is just to have a starting point that works on all targets, since the current driver has issues on some STM32s. Thanks for review; I've added you first to the CODEOWNERS for this driver and rebased.

**Comment by pfalcon** [comment]
Anything precluding this from being merged? @tbursztyka , @nashif , can you please have a look?

**Comment by mbolivar** [comment]
@nashif @galak given that this has the OK from @erwango and @superna9999 I wonder if you could take a look at the change to the SPI core especially in https://github.com/zephyrproject-rtos/zephyr/pull/832/commits/deb3828e9435bf0a3d0f4833521b59815ef06c2a, to see if this is OK to merge.

**Review by tbursztyka — APPROVED** [review]

**Pull request merged**
